docs/hyperparameter_search.rst

```diff
@@ -0,0 +1,136 @@
+    from ray import tune
+    import time
+
+    def training_function(config: Dict):
+        hvd.init()
+        for i in range(config["epochs"]):
+            time.sleep(1)
+            tune.report(test=1, rank=hvd.rank())
+
+    trainable = DistributedTrainableCreator(
+            training_function, num_slots=2, use_gpu=use_gpu)
+    analysis = tune.run(
+            trainable, num_samples=2, config={"epochs": tune.grid_search([1, 2, 3]))
+    print(analysis.best_config)
+
+Not only is the interface very simple, it is also packed with features.
+
+Basic setup
+-----------
+
+Use the `DistributedTrainableCreator`_ function to adapt your Horovod training function to be compatible with Tune.
+
+`DistributedTrainableCreator`_ exposes ``num_hosts``, ``num_slots``, ``use_gpu`` and ``num_cpus_per_slot`` - these parameters allow you to specify the resource allocation of a single "trial" (or "Trainable") which itself can be a distributed training job.
+
+
+.. code-block:: python
+
+    # Each training job will use 2 GPUs.
+    trainable = DistributedTrainableCreator(
+        training_function, num_slots=2, use_gpu=True)
+
+The training function itself must do three things:
+
+1. It must adhere to the Tune Function API signature.
+1. Its body must include a ``horovod.init()`` call.
+1. It must call ``tune.report`` during training, typically called iteratively at the end of every epoch.
+
+Setting up a tuning cluster
+---------------------------
+
+You can easily leverage Ray Tune with Horovod on a laptop, single machine with multiple GPUs, or across multiple machines. To run on a single machine, you can execute your python script as-is (for example, horovod_simple.py, assuming Ray and Horovod are installed properly):
+
+.. code-block:: bash
+
+    python horovod_simple.py
+
+
+To leverage a distributed hyperparameter tuning setup with Ray Tune + Horovod, you’ll need to install Ray and set up a Ray cluster. Ray clusters are started with the Ray Cluster Launcher or manually.
+
+Below, we’ll use the Ray cluster launcher, but you can start Ray on any list of nodes, on any cluster manager or cloud provider.
+
+To use the cluster launcher, you’ll first want to specify a configuration file. Below we have an example of using AWS EC2, but you can easily launch the cluster on any cloud provider:
+
+.. code-block:: yaml
+
+    # ray_cluster.yaml
+    cluster_name: horovod-cluster
+    provider: {type: aws, region: us-west-2}
+    auth: {ssh_user: ubuntu}
+    min_workers: 3
+    max_workers: 3
+
+    # Deep Learning AMI (Ubuntu) Version 21.0
+    head_node: {InstanceType: p3.2xlarge, ImageId: ami-0b294f219d14e6a82}
+    worker_nodes: {InstanceType: p3.2xlarge, ImageId: ami-0b294f219d14e6a82}
+    setup_commands: # Set up each node.
+        - HOROVOD_WITH_GLOO=1 HOROVOD_GPU_OPERATIONS=NCCL pip install horovod[ray]
+
+You can then run ray up ray_cluster.yaml, and a cluster of 4 nodes will be automatically started for you with Ray.
```

Comment:
Is ray up ray_cluster.yaml the command? If so, recommend making this:
`ray up ray_cluster.yaml`

Reply:
done!